Applications/Spire/Source/KeyBindings/TagNameBox.cpp

```diff
@@ -0,0 +1,196 @@
+#include "Spire/KeyBindings/TagNameBox.hpp"
+#include "Spire/KeyBindings/OrderFieldAdditionalTag.hpp"
+#include "Spire/Spire/ListIndexValueModel.hpp"
+#include "Spire/Spire/LocalValueModel.hpp"
+#include "Spire/Styles/Stylist.hpp"
+#include "Spire/Ui/DropDownBox.hpp"
+#include "Spire/Ui/Layouts.hpp"
+
+using namespace boost;
+using namespace boost::signals2;
+using namespace Spire;
+using namespace Spire::Styles;
+
+namespace {
+  struct AdditionalTagNameListModel : ListModel<QString> {
+    std::shared_ptr<ListModel<std::shared_ptr<AdditionalTag>>> m_source;
+    ListModelTransactionLog<QString> m_transaction;
+    scoped_connection m_connection;
+
+    explicit AdditionalTagNameListModel(
+      std::shared_ptr<ListModel<std::shared_ptr<AdditionalTag>>> source)
+      : m_source(std::move(source)),
+        m_connection(m_source->connect_operation_signal(
+          std::bind_front(&AdditionalTagNameListModel::on_operation, this))) {}
+
+    int get_size() const override {
+      return m_source->get_size();
+    }
+
+    const QString& get(int index) const override {
+      return m_source->get(index)->get_name();
+    }
+
+    connection connect_operation_signal(
+        const OperationSignal::slot_type& slot) const override {
+      return m_transaction.connect_operation_signal(slot);
+    }
+
+    void transact(const std::function<void ()>& transaction) override {
+      m_transaction.transact(transaction);
+    }
+
+    void on_operation(
+        const ListModel<std::shared_ptr<AdditionalTag>>::Operation& operation) {
+      visit(operation,
+        [&] (const ListModel<std::shared_ptr<AdditionalTag>>::AddOperation&
+            operation) {
+          m_transaction.push(AddOperation(operation.m_index,
+            operation.get_value()->get_name()));
+        },
+        [&] (const ListModel<std::shared_ptr<AdditionalTag>>::MoveOperation&
+            operation) {
+          m_transaction.push(MoveOperation(operation.m_source,
+            operation.m_destination));
+        },
+        [&] (const ListModel<std::shared_ptr<AdditionalTag>>::RemoveOperation&
+            operation) {
+          m_transaction.push(RemoveOperation(operation.m_index,
+            operation.get_value()->get_name()));
+        },
+        [&] (const ListModel<std::shared_ptr<AdditionalTag>>::UpdateOperation&
+            operation) {
+          m_transaction.push(UpdateOperation(operation.m_index,
+            operation.get_previous()->get_name(),
+            operation.get_value()->get_name()));
+        });
+    }
+  };
+
+  struct AdditionalTagNameValueModel : ValueModel<QString> {
```

Comment:
What's wrong with using a transform?

    make_transform_value_model(std::move(source),
      [] (const auto& tag) {
        return tag->get_name();
      },
      [list] (const auto& name) {
        auto i = std::find_if(list->begin(), list->end(),
          [&] (const auto& tag) {
            return name == tag->get_name();
          });
          if(i == list->end()) {
            throw std::invalid_argument("Tag name not found.");
          }
          return *i;
      });